### ISSUE TITLE
Application process: Force users to fill out their profiles

### ISSUE THREAD

**Issue by mkalininait** [rails-girls-summer-of-code/rgsoc-teams#593]
In 2016 we had 2 problems:
 - Students could submit an application, while for some team members important information was missing (email, location etc)
 - We didn't have enough information about coaches, while they play a crucial role in the program.

**Possible solution:**  
1. After Users sign in with their Github profiles for the first time, they should be redirected to the Edit User Profile form pre-filled with the information from Github. This sign up process should be considered completed only when Users submit this form (with all required fields filled out).  
2. When a user has been added to a team, send an email asking to review/verify their profiles and confirm the membership.  
3. On application submission, validate that all members of the team have filled out all the required fields in their profiles and that they are confirmed team members of their team.  
4. User Profile:
 - Make the following fields required in the User Profile: Name, Location (Country, City)
 - Consider changing the Location fields so that they are consistent with [this requirement](https://github.com/rails-girls-summer-of-code/rgsoc-teams/issues/366) 
 - Make "Github handle" not editable

To be discussed:

 - Add the following fields (all required?): 
  -- "Technologies of expertise" (or "Technologies of interest"? or both?) - tags separated by comma (similar to the "Tags" field in Project)
  -- "Coaching experience" (only for coaches) - multiline textarea field
  -- "Availability (hours per week)" (only for coaches) - numeric field

Please share your opinions on this.

**Comment by carpodaster** [comment]
(The Star Wars kid in me got really excited when I read _Force Users_ 😉 )

![](https://images.duckduckgo.com/iu/?u=https%3A%2F%2Fi.ytimg.com%2Fvi%2F000U_VNRW8s%2Fhqdefault.jpg&f=1)

**Comment by bitboxer** [comment]
For 1: What does "only considered complete" mean? Is there a penalty if a users doesn't fill out the form and closes that window?
For 2: What should the wording of the email be?
For 3: What should the message be if a team member hasn't filled out all required fields?
For 4: When I am seeing this correctly, the github profile name is not editable already, or am I missing something?

**Comment by alicetragedy** [comment]
Intro: The step of signing in with a github handle creates a new user in the db and currently allows that signed in user to e.g. create a team. The idea was to have a two-step process: first step, signing in, just requires the user to accept the oauth via github: this would give a user some sort of “read-only” access for anything that isn't their user profile (i.e. they can edit their user profile, but cannot do anything else like creating teams or applications). Only once the user has confirmed with their email address can they have access to everything else a user should be able to do, like creating teams and so on. I'm pretty sure there are situations where this scenario would fail, and it might be that we haven't thought everything through (actually I'm pretty sure we haven't). Would be happy about improvements on this thought process.

1. If the window is closed and the profile is not fully filled out, the user will be able to view things a normal user can view, but not be able to e.g. create teams (see explanation in intro). This would be the only “penalty”. 

2. I would actually already send this email once the profile has been saved with all required fields filled out. The wording could be something like, 

“Hey!
You've recently logged into the RGSoC teams app with your GitHub account and created a profile. In order to get the best out of the Teams app (like creating a team or submitting a project), we'll need to confirm your email address. To confirm your email, click here (link with token). 

Thanks,
The RGSoC Team"

3. Going along with the scenario described in the intro, the profile would show a fixed notice at the top until the required fields are filled out, something like: “Please fill out all required fields below to complete and verify your profile.” I'd still like it to be more specific (what is the verifying for?) but can't find exact wording for now. 

4. You're correct, I also cannot edit my Github handle currently, but Maria tested it recently and she could, so this might have been fixed in the meanwhile. We'll investigate.


I think this feature might not be fully 100% thought through, and I'm worried that you might encounter some issues on the way. I'd suggest to start thinking about how this could be done, and if you see a huge problem, just ping me or comment in here. I'm happy to help out define some of these steps or the process for you together with Maria, and to make some changes according to your developer's opinion :) Unfortunately I'm not 100% confident about how all of the Teams app works, but we can also talk to Carsten and Max to see what they think of the implementation.

**Comment by alicetragedy** [comment]
Also relevant for 3:

A non-confirmed user can be added to a team, but we should probably wrote a similar notice somewhere to point out that on of the team users has a non-confirmed account.

**Comment by bitboxer** [comment]
Thanks @alicetragedy ! This clears up a lot of things. Will start working on it this afternoon and report back when I have any questions.

**Comment by F3PiX** [comment]
It is very good to strive for complete profiles. OTOH we want to keep the user experience as pleasant as possible. As a aspiring student, I'd want to check out things without telling you everything about myself. 
That said:
- Why don't we grab the emailadresses from GitHub? IIRC it only requires a different permission setting.
- Devise has a cool module :confirmable. It will send a confirmation email, and we can give a user temporary access until the confirmation is required. It is customizable, and I am a big fan, esp because of the user friendlyness, both for the new users and the admins. 
I suppose it is equally easy to email unconfirmed coaches with it.

**Comment by alicetragedy** [comment]
@F3PiX we are grabbing the email addresses from GitHub — unfortunately, since some people have set their email address to private, GH doesn't allow us to access these email addresses directly, which causes the whole problem we have with email addresses missing currently ;)

**Comment by bitboxer** [comment]
@F3PiX the `:confirmable` will be used to confirm the email address from the user 👍

**Comment by F3PiX** [comment]
Aha, of course, sorry.

About the github handle: I think PR #514 handled the non-editability 💭

**Comment by bitboxer** [comment]
Okay, will ignore the non editable in this ticket. If it is still an issue, a new ticket should be created that addresses this issue.

**Comment by bitboxer** [comment]
Current status: will finish this ticket tomorrow and need feedback on the "to be discussed" section till tomorrow afternoon 👍

**Comment by alicetragedy** [comment]
awesome! 🎉

**Comment by mkalininait** [comment]
@bitboxer We will make a decision on the "to be discussed" section by tomorrow 14 CEST ish. Hope it works for you!

**Comment by bitboxer** [comment]
Of course :). I can always delay this and do other things.

**Comment by mkalininait** [comment]
@bitboxer @alicetragedy I would like to leave some of my thoughts on this issue in addition to what has been discussed above.
The logic of creating users/teams/applications is becoming more complicated, and it's important to try to keep things transparent for users.

For example:
1. A user can't do certain things until he/she fills out his/her profile and confirms his/her email. It would be good to make it clear why one can't, say, create a team to avoid messages to our support: "I can't create a team, what am I doing wrong?". Maybe we should avoid hiding UI elements, but instead throw an error message on pressing buttons.
2. "The profile would show a fixed notice at the top until the required fields are filled out" + "we should probably wrote a similar notice somewhere to point out that on of the team users has a non-confirmed account" <-- in addition to that, we should probably write the same thing on the Application Form, so that users don't get surprised at the error when they submit the form.
3. We're gonna send an email once a profile has been saved (to confirm an email address). Should we try to send an email to team members once someone added them to a team (to ask them fill out their profiles)? Otherwise how should they learn about it rather than from the person who added them? There is no guarantee that we have a user email from Github, but some of them will be available at least.

Sorry if I'm adding too much stuff into this issue. Don't want to overwhelm anyone :)

**Comment by bitboxer** [comment]
For 1: If you click on apply now without a confirmed account, you get a warning saying you need to confirm it, if you confirmed it and need to create a team, you are shown the team create page with an explanation that you need to do that.
2. Yes
3. Can you create a new ticket for this? This ticket is getting too big for my taste and I would love to prevent to add more features to it.

**Comment by bitboxer** [comment]
And would it be possible to also create a new ticket for the things in the "to be discussed" section? That way I can finish this ticket and we don't have a huge pull requests with tons of feature in it.

**Comment by bitboxer** [comment]
While we are at it: the change of the location input in the user form could also be another ticket because it doesn't touch any of the other things I am doing here.

**Comment by mkalininait** [comment]
@bitboxer Ok, I'll create a few more issues 👍

**Comment by bitboxer** [comment]
Awesome. That is a huge relieve for me. This pull request will already be relatively huge with changes all over the place 👍

**Comment by bitboxer** [comment]
Created a PR for this. Hope I haven't forgotten something important

**Comment by carpodaster** [comment]
@mkalininait I'm assigning this to you for verification. @bitboxer's PR has been merged.

**Comment by mkalininait** [comment]
Testing results:

1\. "If the window is closed and the profile is not fully filled out, the user will be able to view things a normal user can view, but not be able to e.g. create teams"  <- **seems to work if I create my account myself by signing in with Github**
 * **BUT!! If someone creates my account in the Teams app by adding me to a team, then I can sign in and edit my team and fill out my application without filling out my profile first.** Bug #629 is a result on this.
 * **Also, once I saved my profile and verified my email, I can change the email right afterwards, not confirm it, and still have access to everything like a verified user.**

2\. "I would actually already send this email once the profile has been saved with all required fields filled out."
 * **Seems to work to verify the first email. Once I change my verified email to something else, I don't get the confirmation email and still have full access to the Teams app** (see item 1 in this list).

3\. Notices in UI.

3a. "the profile would show a fixed notice at the top until the required fields are filled out, something like: “Please fill out all required fields below to complete and verify your profile.”" 
 * **We can see this message on the "Edit profile" form, but not on the Profile page. I thought it would be useful to show it on the Profile page. What did you mean @alicetragedy ?**

![screen shot 2017-01-31 at 01 14 20](https://cloud.githubusercontent.com/assets/1745735/22447470/fa9b548e-e752-11e6-97a9-748cdfc399d4.png)

3b. "A non-confirmed user can be added to a team, but we should probably wrote a similar notice somewhere to point out that on of the team users has a non-confirmed account."
 * **NOT implemented**

3c. "we should probably write the same thing on the Application Form, so that users don't get surprised at the error when they submit the form."
 * **Instead of the Application form, we show the message: "You need to click on the link in the email to confirm your account before you can create an application." Sometimes this message is not relevant, for example when I signed in with my Github, but haven't saved it yet and therefore didn't get any emails yet. I'd better write something like: "You need to fill out your account and confirm your email before you can create an application." Also: we don't show any notices, if one of my team members (not me) had his/her profile not filled out.**

![screen shot 2017-01-31 at 01 30 36](https://cloud.githubusercontent.com/assets/1745735/22447851/66dcf0d8-e755-11e6-9912-9fb0207e3ff6.png)

4\. "if you confirmed it [profile] and need to create a team, you are shown the team create page with an explanation that you need to do that."
 * **If I have my profile filled out, but don't have a team yet, then once I click on "Apply now", I see two notices saying the same thing. I think we can delete one.**
![screen shot 2017-01-31 at 01 39 26](https://cloud.githubusercontent.com/assets/1745735/22447992/6c98003e-e756-11e6-80b7-7460d3655b25.png)

5\. "A user can't do certain things until he/she fills out his/her profile and confirms his/her email. [...] Maybe we should avoid hiding UI elements, but instead throw an error message on pressing buttons."
 * **NOT implemented. I don't see the buttons to create a new team or a new project. They are hidden. It's not critical though.**

**Comment by alicetragedy** [comment]
@mkalininait not sure if this will still be relevant after Carsten's upcoming PR, but yeah, I think for 3a it would make a lot of sense to have a general notice at the top of the profile page maybe to show that the profile is still missing something — or anywhere where a filled out profile is required to do something? I'm not sure about the difficulty of implementation there. :(

**Comment by carpodaster** [comment]
Thank you, @mkalininait, for your excessive testing! I took the liberty of tossing in a few headlines to be better visualize the different items.

**Comment by mkalininait** [comment]
I guess we still need to discuss the "To be discussed" part of the issue? :)

**Comment by mkalininait** [comment]
I'd like to close this task. All the issues from it have been either implemented or moved to separate issues (#609, #612, #618, #619). Any objections?
On another note, I remember that we wanted to rework the entire sign up / sign in system before the next season. Is it still the plan @carpodaster ?

**Comment by carpodaster** [comment]
>  I remember that we wanted to rework the entire sign up / sign in system before the next season

@mkalininait yes, that is still the plan. Haven't poured any further thoughts into this, though, and it's probably more of a UX thing than actual programming.

But aye, let's close the task. Will you do the honours, Marias? :)

**Comment by mkalininait** [comment]
@carpodaster My pleasure! :)
